---
title: "Temp_Iso_Turnover"
author: "Aileen Lavelle"
date: "2024-06-04"
output:
  word_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r Setup}
#install.packages(c("ape", "phytools", "readxl", "dplyr", "rotl", "officer", "flextable"))
#install.packages("MCMCglmm", "phylosignal")
library(ape)
library(phytools)
library(readxl)
library(dplyr)
library(rotl)
library(officer)
library(flextable)
library(httr)
library(jsonlite)
library(ggplot2)
library(nlme)
library(janitor)
library(MCMCglmm)
library(phylosignal)
library(paletteer)
library(MuMIn)
library(phylolm)

```

```{r Data Cleaning}
# Read the Excel file
data <- read_excel("~/Desktop/UF/Temp_Iso_Turnover/Main Dataset.xlsx")

# Removing parentheses and asterisks  
cleaned_data <- data %>%
  mutate(`Scientific name` = gsub("[()]", "", `Scientific name`)) %>%
  mutate(`Scientific name` = trimws(`Scientific name`))

cleaned_data <- cleaned_data %>%
  mutate(`Body Temp (°C)` = as.numeric(gsub("\\*", "", `Body Temp (°C)`)),
         `Temp (°C)` = as.numeric(gsub("\\*", "", `Temp (°C)`))) %>%
  mutate(Temperature = ifelse(!is.na(`Body Temp (°C)`), `Body Temp (°C)`, `Temp (°C)`))

#selecting relevant columns
cleaned_data <- cleaned_data %>%
  select(`Scientific name`, `Endo-/Ectothermy`, `Isotope`, `Tissue`, Temperature, `Mass (g)`, `Half-life (days)`)

#Some of the tissues are capitalized & others aren't, to filter they must be the same
#making all lowercase
cleaned_data <- cleaned_data %>%
  mutate(Tissue = tolower(Tissue))

#ensure no repeat tissue types
unique(cleaned_data$Tissue)
#yay :)

# Filter data for each isotope
isotope_13C <- cleaned_data %>%
  filter(Isotope == "13C")

isotope_15N <- cleaned_data %>%
  filter(Isotope == "15N")

isotope_34S <- cleaned_data %>%
  filter(Isotope == "34S")

```

```{r Fixing Naming Discrepencies for 13C dataset}
#I run into some issues later on from naming discrepencies, so I am fixing here
#"Pleuronectes americanus" &  "Pseudopleuronectes americanus" are both listed but according to NOAA report from 1992, this species is now refered to by "Pleuronectes americanus" but is only in open tree of life as "Pseudopleuronectes americanus" so I am going to go with "Pseudopleuronectes americanus" & remove the other name
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Pleuronectes americanus", 
                                    "Pseudopleuronectes americanus", 
                                    `Scientific name`))

#"Bufo_calamita" is not the correct nomenclature. Bufonidae is the family but the correct species name is Epidalea calamita, I am going to change this
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Bufo calamita", 
                                    "Epidalea calamita", 
                                    `Scientific name`))
#Unresolved issues: These two are subspecies, not species; Calidris alpina pacifica, Ursus arctos yesoyensis
##I am changing the subspecies to species for ID matching purposes
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Calidris alpina pacifica",
                                    "Calidris alpina", 
                                    `Scientific name`))
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Ursus arctos yesoyensis",
                                    "Ursus arctos", 
                                    `Scientific name`))

#"catharacta skua" is not the correct nomenclature; it is Stercorarius skua since 1764.
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Catharacta skua", 
                                    "Stercorarius skua", 
                                    `Scientific name`))

#"Crassostrea gigas" is not the correct nomenclature; it is Magallana gigas since 2017
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Crassostrea gigas", 
                                    "Magallana_gigas", 
                                    `Scientific name`))
#"Dendroica coronata" is listed in the dataset (Linnaeus, 1766) but is now classified as "Setophaga coronata"
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Dendroica coronata", 
                                    "Setophaga coronata", 
                                    `Scientific name`))

#"lama pacos" is "Vicugna pacos" in Open Tree of Life
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Lama pacos", 
                                    "Vicugna pacos", 
                                    `Scientific name`))
#"elaphe guttata" in dataset is "Pantherophis guttatus" in Open Tree of Life
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Elaphe guttata", 
                                    "Pantherophis guttatus", 
                                    `Scientific name`))
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Elaphe guttata guttata", 
                                    "Pantherophis guttatus", 
                                    `Scientific name`))
#"nereis virens" is an older name for "Alitta virens"
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Nereis virens", 
                                    "Alitta virens", 
                                    `Scientific name`))
#"litopenaeus vannamei" is "Penaeus vannamei" in Open Tree of Life
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Litopenaeus vannamei", 
                                    "Penaeus vannamei", 
                                    `Scientific name`))

#"Trichechus manatus latirostris" is "Trichechus manatus" in Open Tree of Life
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Trichechus manatus latirostris", 
                                    "Trichechus manatus", 
                                    `Scientific name`))

#"gadus morhua" has "(in_domain_Bacteria)" which is likely messing up the search string
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Gadus morhua", 
                                    "Gadus_morhua_(in_domain_Bacteria)", 
                                    `Scientific name`))
#"Terrapene_ornate luteola" should be Terrapene ornata luteola
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Terrapene ornate luteola",
                                    "Terrapene_ornata_luteola", 
                                    `Scientific name`))
#"Terrapene_ornata luteola" should be Terrapene ornata luteola
isotope_13C <- isotope_13C %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Terrapene ornata luteola",
                                    "Terrapene_ornata_luteola", 
                                    `Scientific name`))

# Replace species names in isotope_13C with the correct format
isotope_13C$`Scientific name` <- sub(" ", "_", isotope_13C$`Scientific name`)
# Check the cleaned 13C species names
unique_species_names <- unique(isotope_13C$`Scientific name`)
unique_species_names
```

```{r Query Open Tree of life to create tree for 13C}
# Match species names with the Open Tree of Life taxonomy
resolved_names <- tnrs_match_names(names = unique_species_names)

# Identify unmatched species
unmatched <- resolved_names[is.na(resolved_names$ott_id), ]
if (nrow(unmatched) > 0) {
  print("Unmatched species:")
  print(unmatched$search_string)
}

#Unresolved issues: these two are genuses, not species. Milacidae spp., Rhinogobius spp.
##I am removing "Milacidae spp." and "Rhinogobius spp." for now
unique_species_names <- unique_species_names[!grepl("Milacidae spp.|Rhinogobius spp.", unique_species_names)]

#Unresolved issues: These two are subspecies, not species; Calidris alpina pacifica, Ursus arctos yesoyensis
##I am changing the subspecies to species for ID matching purposes
unique_species_names[unique_species_names == "Calidris alpina pacifica"] <- "Calidris alpina"
unique_species_names[unique_species_names == "Ursus arctos yesoyensis"] <- "Ursus arctos"

resolved_names <- tnrs_match_names(names = unique_species_names)

# Remove unmatched species from the resolved names
resolved_names <- resolved_names[!is.na(resolved_names$ott_id), ]

# Get the Open Tree of Life IDs for the matched names
ott_ids <- resolved_names$ott_id
# Check the resolved OTT IDs
resolved_names

# ID 378311 has the following flag: hidden_inherited - descends from node flagged hidden
# ID 1096146 has the following flag: incertae_sedis - in source taxonomy, was a member of an "incertae sedis" container (also "unallocated", "unclassified", "mitosporic")
# ID 7524264 has the following flags :not_otu - the name suggests that this is not a taxon. Keywords interpreted this way include "uncultured", "unclassified", "unidentified", "unknown", "metagenome", "other sequences", "artificial", "libraries", "tranposons", and a few others. Also "sp." when at the end of a name. Also, any node descended from such a node. This flag is applied to NCBI taxa but not to SILVA taxa & incertae_sedis - in source taxonomy, was a member of an "incertae sedis" container (also "unallocated", "unclassified", "mitosporic")
# ID 935410 has the following flag: hidden_inherited - descends from node flagged hidden
## I am going to remove these for now 
ott_ids <- ott_ids[!ott_ids %in% c(378311, 1096146, 7524264, 935410)]

# Remove the specified OTT IDs from resolved_names
resolved_names <- resolved_names[!(resolved_names$ott_id %in% c(378311, 1096146, 7524264, 935410)), ]

# Retrieve the phylogenetic tree
phylo_tree <- tol_induced_subtree(ott_ids = ott_ids)

# Plot the tree
plot(phylo_tree, main = "Phylogenetic Tree from Open Tree of Life")
# Check the class of pruned_tree
class(phylo_tree)
summary(phylo_tree)
```

```{r Adding Branch Lengths for 13C}
#The Grafen method is applied when the phylogentic tree does not yet have branch lengths. The Grafen method assumes that the amount of evolutionary change is proportional to the time elapsed, and it standardizes the phylogenetic tree so that the distances from the root to each tip are equal.

# Assign branch lengths using the Grafen method
phylo_tree_with_brlen <- compute.brlen(phylo_tree, method = "Grafen")

# Plot the tree with branch lengths
plot(phylo_tree_with_brlen)

#Ensure branch lenghts are included
print(phylo_tree_with_brlen)
phylo_brlen_names <- phylo_tree_with_brlen$tip.label
phylo_brlen_names
#yay :)
```

```{r Check Filtering for 13C}
#Phylogenetic tree tip names with the assigned branch lengths
base_names_phylo <- phylo_tree_with_brlen$tip.label
base_names_phylo
# Assuming base_names_phylo is a list of full names
scientific_names_phylo <- sapply(strsplit(base_names_phylo, "_ott"), `[`, 1)
phylo_df <- data.frame(
  scientific_name = scientific_names_phylo,
  branch_name = base_names_phylo,  # Full names like "Mus_musculus_ott542509"
  stringsAsFactors = FALSE
)
new_dataset_df <- merge(new_dataset_df, phylo_df, by.x = "scientific_name", by.y = "scientific_name", all.x = TRUE)

#write.csv(new_dataset_df, file = "~/Desktop/UF/Temp_Iso_Turnover/iso_13C_filtered.csv", row.names = FALSE)
iso_13C_filtered <- read.csv( "~/Desktop/UF/Temp_Iso_Turnover/iso_13C_filtered.csv")

#what is the range of temperatures for ectotherms
temperature_ranges <- isotope_13C %>%
  filter(!is.na(Temperature)) %>%
  filter(`Endo-/Ectothermy` == "ecto-") %>%
  group_by(`Scientific name`) %>%
  summarise(min_temperature = min(Temperature, na.rm = TRUE),
            max_temperature = max(Temperature, na.rm = TRUE),
            temp_range = max_temperature - min_temperature)

# View the temperature ranges
temperature_ranges

ggplot(temperature_ranges, aes(x = `Scientific name`, ymin = min_temperature, ymax = max_temperature)) +
  geom_linerange() +
  geom_point(aes(y = min_temperature), color = "#FFC125", size = 3) +
  geom_point(aes(y = max_temperature), color = "#FF8247", size = 3) +
  coord_flip() +
  labs(title = "Temperature Ranges for Ectotherms",
       x = "Scientific Name",
       y = "Temperature (°C)") +
  theme_minimal()
#now that we have ensured we have the same 
# Clean the column names of your dataset
iso_13C_filtered <- iso_13C_filtered %>%
  clean_names()

```

```{r Perform Phylogenetic Generalized Least Squares (PGLS) for 13C}
#PGLS incorporates phylogenetic information into the regression analysis by using a covariance matrix that reflects the evolutionary relationships among species. This allows for more accurate estimation of regression parameters and hypothesis testing.

#is there a correlation btw body temperature & turnover rate for each tissue type?
# Plotting the correlation between body temperature and turnover rate for each tissue type
ggplot(iso_13C_filtered, aes(x = temperature, y = half_life_days)) +
  geom_point(aes(color = tissue)) +
  geom_smooth(method = "lm", se = FALSE, aes(color = tissue)) +
  labs(title = "Correlation between Body Temperature and Turnover Rate of 13C",
       x = "Body Temperature (°C)",
       y = "Turnover Rate (Half-life in days)",
       color = "Tissue Type") +
  scale_color_paletteer_d("colorBlindness::Blue2DarkOrange18Steps") +
  theme_minimal()

iso_13C_filtered <- read.csv( "~/Desktop/UF/Temp_Iso_Turnover/iso_13C_filtered.csv")

# Create new columns for the transformations
iso_13C_filtered$ln_turnover_time <- log(iso_13C_filtered$half_life_days)
iso_13C_filtered$ln_body_mass <- log(iso_13C_filtered$mass_g)

# Check for missing values in the relevant columns
missing_values <- iso_13C_filtered  %>%
  select(scientific_name, temperature, ln_body_mass, ln_turnover_time) %>%
  summarise_all(funs(sum(is.na(.))))

print(missing_values)

# Remove rows with missing values
iso_13C_filtered <- iso_13C_filtered %>%
  filter(!is.na(scientific_name) & !is.na(temperature) & !is.na(ln_body_mass) & !is.na(ln_turnover_time))
# Fit the GLS model with the corPagel correlation structure to account for multiple points per species


pglsModel_13C <- gls(ln_turnover_time ~ ln_body_mass * temperature, 
                     correlation = corPagel(1, phy = phylo_tree_with_brlen, form = ~ branch_name),
                     data = iso_13C_filtered, method = "ML")


### model comparison between different evolutionary mechanisms
rownames(iso_13C_filtered) <- iso_13C_filtered$branch_name

### null model
NULLpglsModel_13C <- gls(ln_turnover_time ~ ln_body_mass * temperature,
                         data = iso_13C_filtered, method = "ML")
### brownian motion
BMpglsModel_13C  <- phylolm::phylolm(ln_turnover_time ~ ln_body_mass * temperature,
                         data = iso_13C_filtered, phy = phylo_tree_with_brlen, model="BM" )

### pagels lambda
### OU process
### bloombergs K






# Summary of the PGLS model
summary(pglsModel_13C)
# Calculating the R²
rse <- 0.182732
df_residual <- 435
# Calculate the Residual Sum of Squares (SSR)
ssr <- (rse^2) * df_residual
# Calculate the Total Sum of Squares (SST) 
sst <- var(iso_13C_filtered$ln_turnover_time) * (length(iso_13C_filtered$ln_turnover_time) - 1)
# Calculate R²
r_squared <- 1 - (ssr / sst)
cat("R² for pglsModel_13C:", r_squared, "\n")

# Calculate the transformed Y values
iso_13C_filtered$ln_turnover_mass_rse <- iso_13C_filtered$ln_turnover_time / ((iso_13C_filtered$ln_body_mass) ^ rse)
slope_13C <- coef(pglsModel_13C)["temperature"]
plot(iso_13C_filtered$temperature, iso_13C_filtered$ln_turnover_mass_rse,
     xlab = "Temperature (°C)", 
     ylab = "ln(Turnover Time / Mass^(rse))",
     main = paste("Temperature dependence of 13C turnover time\naccounting for body mass & phylogeny\nSlope =", round(slope_13C, 3)),
     col = "black", pch = 16)
abline(a = coef(pglsModel_13C)[1], b = slope_13C, col = "#AB82FF")
legend("topright", legend = "PGLS Regression Line", lty = 1, col = "#AB82FF")

# Filter the dataset to include only liver tissue data
liver_data <- iso_13C_filtered %>%
  filter(tissue == "liver")
# Remove rows with NA or Inf values
liver_data <- liver_data[complete.cases(liver_data) & !is.infinite(liver_data$ln_turnover_time) &
                                !is.infinite(liver_data$temperature) &
                                !is.infinite(liver_data$ln_body_mass), ]
# Fit the GLS model with the corBrownian correlation structure for liver tissue
pglsModel_liver <- gls(ln_turnover_time ~ temperature + ln_body_mass,
                       correlation = corPagel(1, phy = phylo_tree_with_brlen, form = ~ branch_name),
                       data = liver_data, method = "ML")

# Summary of the PGLS model
summary(pglsModel_liver)
rse <- 0.915523  # Residual standard error from the model summary
df_residual <- 47  # Residual degrees of freedom
# Calculate the Residual Sum of Squares (SSR)
ssr <- (rse^2) * df_residual
# Calculate the Total Sum of Squares (SST) from the observed data
sst <- var(liver_data$normalized_turnover_time) * (length(liver_data$normalized_turnover_time) - 1)
# Calculate R²
r_squared <- 1 - (ssr / sst)
cat("R² for pglsModel_liver:", r_squared, "\n")

# Extract the slope of the regression line
slope_liver <- coef(pglsModel_liver)[2]

# Plotting the transformed dependent variable against temperature for liver tissue
plot(liver_data$temperature, liver_data$normalized_turnover_time,
     xlab = "Temperature (°C)", ylab = "ln(Turnover Time / Mass^(1/4))",
     main = paste("Temperature dependence of 13C turnover time in liver tissue\naccounting for body mass & phylogeny\nSlope =", round(slope_liver, 3)))
abline(a = coef(pglsModel_liver)[1], b = coef(pglsModel_liver)[2], col = "#836FFF")
legend("topright", legend = "PGLS Regression Line", lty = 1, col = "#836FFF")

# Filter the dataset to include only plasma tissue data
plasma_data <- cleaned_dataset_df %>%
  filter(tissue == "plasma")
# Fit the GLS model with corPagel for plasma tissue
pglsModel_plasma <- gls(normalized_turnover_time ~ temperature + ln_body_mass,
                        correlation = corBrownian(phy = phylo_tree_with_brlen, form = ~ scientific_name),
                        data = plasma_data, method = "ML")
# Summary of the PGLS model
summary(pglsModel_plasma)
rse <- 0.7739796
df_residual <- 31
# Calculate the Residual Sum of Squares (SSR)
ssr <- (rse^2) * df_residual

# Calculate the Total Sum of Squares (SST) from the observed data in plasma_data
sst <- var(plasma_data$normalized_turnover_time) * (length(plasma_data$normalized_turnover_time) - 1)

# Calculate R²
r_squared <- 1 - (ssr / sst)
cat("R² for pglsModel_plasma:", r_squared, "\n")

# Extract the slope of the regression line
slope_plasma <- coef(pglsModel_plasma)[2]
# Plotting the transformed dependent variable against temperature for liver tissue
plot(plasma_data$temperature, plasma_data$normalized_turnover_time,
     xlab = "Temperature (°C)", ylab = "ln(Turnover Time / Mass^(1/4))",
     main = paste("Temperature dependence of 13C turnover time in plasma tissue\naccounting for body mass & phylogeny\nSlope =", round(slope_liver, 3)))
abline(a = coef(pglsModel_plasma)[1], b = coef(pglsModel_plasma)[2], col = "#EE5C42")
legend("topright", legend = "PGLS Regression Line", lty = 1, col = "#EE5C42")

```

```{r Insightful Plots for 13C}
#Dr.Gillooly recommended plotting ln(turnover time/M^1/4) for all of the carbon turnover data
# Plot ln(turnover time / M^1/4) for all data
plot(new_dataset_df$temperature, new_dataset_df$normalized_turnover_time,
     xlab = "Average Temperature (°C)", ylab = "ln(Turnover Time / Mass^(1/4))",
     main = "Temperature Dependence of 13C Turnover Time for All Data")

# Add regression line from PGLS model (if available)
plot(new_dataset_df$temperature, new_dataset_df$normalized_turnover_time,
     xlab = "Temperature (°C)", 
     ylab = "ln(Turnover Time / Mass^(1/4))", 
     main = "Temperature Dependence of 13C Turnover Time")
coef_pgls <- coef(pglsModel_13C)
abline(a = coef_pgls[1], b = coef_pgls[2], col = "#836FFF")

# Calculate coordinates for text in the lower left corner
text_x <- min(new_dataset_df$temperature, na.rm = TRUE) + 0.1 * diff(range(new_dataset_df$temperature, na.rm = TRUE))
text_y <- min(new_dataset_df$normalized_turnover_time, na.rm = TRUE) + 0.1 * diff(range(new_dataset_df$normalized_turnover_time, na.rm = TRUE))
regression_text <- sprintf("Slope: %.4f\nIntercept: %.4f", coef_pgls[2], coef_pgls[1])
text(text_x, text_y, labels = regression_text, pos = 4, col = "#836FFF")
legend("topright", legend = "PGLS Regression Line", lty = 1, col = "#836FFF")


# Calculate ln(turnover time / M^1/4) for the complete dataset
new_dataset_df <- new_dataset_df %>%
  mutate(ln_turnover_time = log(half_life_days),
         ln_body_mass = log(mass_g),
         normalized_turnover_time = ln_turnover_time - (1/4) * ln_body_mass)

# Aggregate data by species
averaged_data <- new_dataset_df %>%
  group_by(scientific_name) %>%
  summarize(
    avg_temperature = mean(temperature, na.rm = TRUE),
    avg_normalized_turnover_time = mean(normalized_turnover_time, na.rm = TRUE),
    avg_ln_body_mass = mean(ln_body_mass, na.rm = TRUE)
  )

# Plot ln(turnover time / M^1/4) for all data
plot(new_dataset_df$temperature, new_dataset_df$normalized_turnover_time,
     xlab = "Average Temperature (°C)", ylab = "ln(Turnover Time / Mass^(1/4))",
     main = "Temperature Dependence of 13C Turnover Time",
     pch = 16, col = "grey")

# Overlay the averaged data points
points(averaged_data$avg_temperature, averaged_data$avg_normalized_turnover_time, 
       pch = 19, col = "#00F5FF")
# Fit linear model to averaged data
lm_avg <- lm(avg_normalized_turnover_time ~ avg_temperature, data = averaged_data)
abline(lm_avg, col = "#FF8247")
legend("topright", legend = c("All Data", "Averaged Data", "Averaged Data Regression Line"), 
       pch = c(16, 19, NA), col = c("grey", "#00F5FF", "#FF8247"), lty = c(NA, NA, 1))

# Fit linear model to averaged data
lm_avg <- lm(avg_normalized_turnover_time ~ avg_temperature, data = averaged_data)

# Get the slope (coefficient of avg_temperature)
slope <- coef(lm_avg)["avg_temperature"]

# Plot averaged data points with regression line
plot(averaged_data$avg_temperature, averaged_data$avg_normalized_turnover_time,
     xlab = "Average Temperature (°C)", ylab = "ln(Turnover Time / Mass^(1/4))",
     main = "Averaged Data: Temperature vs ln(Turnover Time / Mass^(1/4))",
     pch = 19, col = "#FF82AB")

# Add regression line
abline(lm_avg, col = "#1E90FF")

# Add legend and slope annotation
legend("topright", legend = paste("Slope =", round(slope, 4)), col = "#1E90FF", lty = 1)

```

```{r MCMCglmm with Species as a Fixed Effect for 13C}
#MCMCglmm with Species as a Fixed Effect
# Check for missing values in fixed predictors

summary(new_dataset_df$temperature)
summary(new_dataset_df$ln_body_mass)
summary(new_dataset_df$normalized_turnover_time)

# Check for missing values in random predictors
table(is.na(new_dataset_df$scientific_name))
# Remove rows with missing values in key columns
cleaned_data <- new_dataset_df %>%
  filter(!is.na(temperature) & !is.na(ln_body_mass) & !is.na(ln_turnover_time) & !is.na(scientific_name))
# Ensure scientific_name is a factor
head(iso_13C_filtered)
iso_13C_filtered$scientific_name <- as.factor(iso_13C_filtered$scientific_name)
iso_13C_filtered <- iso_13C_filtered %>%
  group_by(scientific_name) %>%
  mutate(spec_id = paste0("spec_", row_number()))

# Fit the MCMCglmm model with species as a fixed effect
Ginv <- list(branch_name=inverseA(phylo_tree_with_brlen))$Ainv
head(iso_13C_filtered)

prior <- list(
  R = list(V = 1, nu = 0.0001), 
  G = list(
    G1 = list(V = 1, nu = 0.0001),
    G2 = list(V = 1, nu = 0.0001)
  )
)

mcmc_model <- MCMCglmm(ln_turnover_time ~ temperature * ln_body_mass,
                       prior=prior, random = ~branch_name + spec_id, 
                       ginverse = Ginv, 
                       pedigree = tree,
                       data = iso_13C_filtered, 
                       family = "gaussian", 
                       verbose = FALSE,
                       nitt=20000, thin=10, burnin=10000)

# Check model summary
summary(mcmc_model)
#the lower the DIC, the better the model. Temperature and body mass both significantly influence the normalized turnover time.

lambda <- mcmc_model$VCV[,"branch_name"] / (mcmc_model$VCV[,"branch_name"] + mcmc_model$VCV[,"units"])
mean_lambda <- mean(lambda)
mean_lambda
posterior.mode(lambda)
HPDinterval(lambda)

# Extract mean fixed effects
fixed_effects <- colMeans(mcmc_model$Sol)

#matrix for new data
X <- model.matrix(~ temperature * ln_body_mass, data = new_data)

# Ensure compatibility
stopifnot(ncol(X) == length(fixed_effects))

# Calculate predictions
predictions <- X %*% fixed_effects

# Extract mean fixed effects coefficients from the model
fixed_effects <- colMeans(mcmc_model$Sol)

# Generate a smooth grid for predictions
smooth_data <- data.frame(
  temperature = seq(min(iso_13C_filtered$temperature), max(iso_13C_filtered$temperature), length.out = 100),
  ln_body_mass = rep(mean(iso_13C_filtered$ln_body_mass), 100)  # Fixed value across all rows
)

# Create the design matrix for the smooth data
X_smooth <- model.matrix(~ temperature * ln_body_mass, data = smooth_data)

# Compute predictions for the smooth data
predictions_smooth <- X_smooth %*% fixed_effects

# Manually specify the coefficients and intercept for the line equation
slope_temp <- -0.05
slope_mass <- 0.19
intercept <- 4.17

# Generate a smooth grid for temperature
smooth_temp <- seq(min(iso_13C_filtered$temperature), max(iso_13C_filtered$temperature), length.out = 100)

# Calculate predictions using the specified formula
# Use the mean value of ln_body_mass for simplicity in this example
mean_mass <- mean(iso_13C_filtered$ln_body_mass)
predictions_manual <- slope_temp * smooth_temp + slope_mass * mean_mass + intercept

# Plot actual data points
plot(iso_13C_filtered$temperature, iso_13C_filtered$ln_turnover_time, 
     xlab = "Body Temperature (°C)", 
     ylab = bquote(ln(t[1/2]/m[.(0.19)]^{}~"[days/g"[.(0.19)]^{}~"]")),
     main = paste("Predicted ln(Turnover Time) Based on Lambda =", round(mean_lambda, 3)),
     col = "black", pch = 16)

# Add the manually computed line
lines(smooth_temp, predictions_manual, col = "#FF8247", lwd = 2)

# Add R² and equation to the plot in the lower left corner
text(x = min(iso_13C_filtered$temperature), y = min(iso_13C_filtered$ln_turnover_time) + 0.5,
     labels = bquote(R^2 == .(round(r_squared, 3))),
     pos = 4, cex = 0.8)

text(x = min(iso_13C_filtered$temperature), y = min(iso_13C_filtered$ln_turnover_time),
     labels = bquote("y = " ~ -0.05 * temp + 0.19 * mass + 4.17),
     pos = 4, cex = 0.8)
```

```{r Fixing Naming Discrepencies for 15N dataset}
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Clupea pallasi", 
                                    "Clupea pallasii", 
                                    `Scientific name`))

isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Ctenopharyngodon idellus", 
                                    "Ctenopharyngodon idella", 
                                    `Scientific name`))
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Bufo calamita", 
                                    "Epidalea calamita", 
                                    `Scientific name`))
#"catharacta skua" is not the correct nomenclature; it is Stercorarius skua since 1764.
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Catharacta skua", 
                                    "Stercorarius skua", 
                                    `Scientific name`))
#"Trichechus manatus latirostris" is "Trichechus manatus" in Open Tree of Life
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Trichechus manatus latirostris", 
                                    "Trichechus manatus", 
                                    `Scientific name`))
#Unresolved issues: These two are subspecies, not species; Calidris alpina pacifica
##I am changing the subspecies to species for ID matching purposes
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Calidris alpina pacifica",
                                    "Calidris alpina", 
                                    `Scientific name`))
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Elaphe guttata guttata", 
                                    "Pantherophis guttatus", 
                                    `Scientific name`))
#"Crassostrea gigas" is not the correct nomenclature; it is Magallana gigas since 2017
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Crassostrea gigas", 
                                    "Magallana_gigas", 
                                    `Scientific name`))
#"Dendroica coronata" is listed in the dataset (Linnaeus, 1766) but is now classified as "Setophaga coronata"
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Dendroica_coronata", 
                                    "Setophaga_coronata", 
                                    `Scientific name`))
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Dendoica_coronata", 
                                    "Setophaga_coronata", 
                                    `Scientific name`))
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Pleuronectes americanus", 
                                    "Pseudopleuronectes americanus", 
                                    `Scientific name`))
#"nereis virens" is an older name for "Alitta virens"
isotope_15N <- isotope_15N %>%
  mutate(`Scientific name` = ifelse(`Scientific name` == "Nereis virens", 
                                    "Alitta virens", 
                                    `Scientific name`))

#Unresolved issues: these two are genuses, not species. Milacidae spp., Rhinogobius spp.
##I am removing "Milacidae spp." and "Rhinogobius spp." for now
# Remove entries for "Milacidae spp." and "Rhinogobius spp." from the isotope_15N dataset
isotope_15N <- isotope_15N[!grepl("Milacidae spp.|Rhinogobius sp.|Elimia sp.|Rhinogobius spp.", isotope_15N$`Scientific name`), ]

#Lavigeria grandis is hidden interted flag on open tree of life so I will remove for now
isotope_15N <- isotope_15N[!grepl("Oncorhynchus mykiss", isotope_15N$`Scientific name`), ]

#Lavigeria grandis is hidden interted flag on open tree of life so I will remove for now
isotope_15N <- isotope_15N[!grepl("Lavigeria grandis", isotope_15N$`Scientific name`), ]

#Lavigeria grandis is hidden interted flag on open tree of life so I will remove for now
isotope_15N <- isotope_15N[!grepl("Tarebia granifera", isotope_15N$`Scientific name`), ]

#Gadus morhua is hidden interted flag on open tree of life so I will remove for now
isotope_15N <- isotope_15N[!grepl("Gadus morhua", isotope_15N$`Scientific name`), ]


# Replace species names in isotope_13C with the correct format
isotope_15N$`Scientific name` <- sub(" ", "_", isotope_15N$`Scientific name`)
# Check the cleaned 13C species names
unique_species_names_15N <- unique(isotope_15N$`Scientific name`)
unique_species_names_15N
```

```{r Query Open Tree of life to create tree for 15N dataset}
# Match species names with the Open Tree of Life taxonomy
resolved_names_n <- tnrs_match_names(names = unique_species_names_15N)

# Identify unmatched species
unmatched <- resolved_names_n[is.na(resolved_names_n$ott_id), ]
if (nrow(unmatched) > 0) {
  print("Unmatched species:")
  print(unmatched$search_string)
}

# Remove unmatched species from the resolved names
resolved_names_n <- resolved_names_n[!is.na(resolved_names_n$ott_id), ]

# Get the Open Tree of Life IDs for the matched names
ott_ids_nit <- resolved_names_n$ott_id
# Check the resolved OTT IDs
resolved_names_n

# Removing problematic ott_ids from the list
problematic_ids <- c(378311, 935410)  # List the ott_ids that caused the error
ott_ids_nit <- ott_ids_nit[!ott_ids_nit %in% problematic_ids]

# Retrieve the phylogenetic tree
phylo_tree_nitro <- tol_induced_subtree(ott_ids = ott_ids_nit)

# Plot the tree
plot(phylo_tree_nitro, main = "Phylogenetic Tree from Open Tree of Life 15N")
# Check the class of pruned_tree
class(phylo_tree_nitro)
summary(phylo_tree_nitro)
```

```{r Adding Branch Lengths for 15N}
#The Grafen method is applied when the phylogentic tree does not yet have branch lengths. The Grafen method assumes that the amount of evolutionary change is proportional to the time elapsed, and it standardizes the phylogenetic tree so that the distances from the root to each tip are equal.

# Assign branch lengths using the Grafen method
phylo_tree_Nitro_brlen <- compute.brlen(phylo_tree_nitro, method = "Grafen")

# Plot the tree with branch lengths
plot(phylo_tree_Nitro_brlen)

#Ensure branch lenghts are included
print(phylo_tree_Nitro_brlen)
phylo_brlen_Nitro_names <- phylo_tree_Nitro_brlen$tip.label
phylo_brlen_Nitro_names
#yay :)
```

```{r Check Filtering for 15N}
#Phylogenetic tree tip names with the assigned branch lengths
# Assuming base_names_phylo is a list of full names
scientific_names_phylo <- sapply(strsplit(phylo_brlen_Nitro_names, "_ott"), `[`, 1)

# Adjust tip labels to match scientific_names_phylo
phylo_tree_Nitro_brlen$tip.label <- gsub("_ott_", "_ott", phylo_tree_Nitro_brlen$tip.label)

phylo_nitro_df <- data.frame(
  scientific_name = scientific_names_phylo,
  branch_name = phylo_brlen_Nitro_names,  # Full names like "Mus_musculus_ott_542509"
  stringsAsFactors = FALSE
)
iso_15N <- merge(isotope_15N, phylo_nitro_df, by.x = "Scientific name", by.y = "scientific_name", all.x = TRUE)
unmatched_species <- iso_15N[is.na(iso_15N$branch_name), "Scientific name"]
print(unmatched_species)
# Add ln_turnover_time column (natural log of Half-life (days))
iso_15N <- iso_15N %>%
  mutate(ln_turnover_time = log(`Half-life (days)`))

# Add ln_body_mass column (natural log of Mass (g))
iso_15N <- iso_15N %>%
  mutate(ln_body_mass = log(`Mass (g)`))

tree_tip_labels <- phylo_tree_nitro$tip.label
print(tree_tip_labels)

#write.csv(iso_15N, file = "~/Desktop/UF/Temp_Iso_Turnover/iso_15N_filtered.csv", row.names = FALSE)
iso_15N_filtered <- read.csv( "~/Desktop/UF/Temp_Iso_Turnover/iso_15N_filtered.csv")
```

```{r MCMCGlmm 15N dataset}
#MCMCglmm with Species as a Fixed Effect
# Check for missing values in fixed predictors

summary(iso_15N_filtered$Temperature)
summary(iso_15N_filtered$ln_body_mass)
summary(iso_15N_filtered$ln_body_mass)

# Check for missing values in random predictors
table(is.na(iso_15N_filtered$scientific_name))
# Remove rows with missing values in key columns
iso_15N_filtered <- iso_15N_filtered %>%
  filter(!is.na(Temperature) & !is.na(ln_body_mass) & !is.na(ln_turnover_time) & !is.na(Scientific.name))
# Ensure scientific_name is a factor
iso_15N_filtered$Scientific.name <- as.factor(iso_15N_filtered$Scientific.name)
iso_15N_filtered<- iso_15N_filtered %>%
  group_by(Scientific.name) %>%
  mutate(spec_id = paste0("spec_", row_number()))


# Fit the MCMCglmm model with species as a fixed effect
Ginv <- list(branch_name=inverseA(phylo_tree_Nitro_brlen))

tree <-phylo_tree_Nitro_brlen

prior <- list(
  R = list(V = 1, nu = 0.0001), 
  G = list(
    G1 = list(V = 1, nu = 0.0001),
    G2 = list(V = 1, nu = 0.0001)
  )
)
# Extract tip labels from the tree
tree_tips <- phylo_tree_Nitro_brlen$tip.label

# Check the species names in your data
data_species <- iso_15N_filtered$spec_id
colnames(iso_15N_filtered)

mcmc_model <- MCMCglmm(ln_turnover_time ~ Temperature * ln_body_mass,
                       prior=prior, random = ~branch_name, 
                       ginverse = Ginv, 
                       pedigree = tree,
                       data = iso_15N_filtered, 
                       family = "gaussian", 
                       verbose = FALSE,
                       nitt=20000, thin=10, burnin=10000)

# Check model summary
summary(mcmc_model)
#the lower the DIC, the better the model. Temperature and body mass both significantly influence the normalized turnover time.

lambda <- mcmc_model$VCV[,"branch_name"] / (mcmc_model$VCV[,"branch_name"] + mcmc_model$VCV[,"units"])
mean_lambda <- mean(lambda)
mean_lambda
posterior.mode(lambda)
HPDinterval(lambda)

# Extract mean fixed effects
fixed_effects <- colMeans(mcmc_model$Sol)

#matrix for new data
X <- model.matrix(~ temperature * ln_body_mass, data = new_data)

# Ensure compatibility
stopifnot(ncol(X) == length(fixed_effects))

# Calculate predictions
predictions <- X %*% fixed_effects

# Extract mean fixed effects coefficients from the model
fixed_effects <- colMeans(mcmc_model$Sol)

# Generate a smooth grid for predictions
smooth_data <- data.frame(
  temperature = seq(min(iso_13C_filtered$temperature), max(iso_13C_filtered$temperature), length.out = 100),
  ln_body_mass = rep(mean(iso_13C_filtered$ln_body_mass), 100)  # Fixed value across all rows
)

# Create the design matrix for the smooth data
X_smooth <- model.matrix(~ temperature * ln_body_mass, data = smooth_data)

# Compute predictions for the smooth data
predictions_smooth <- X_smooth %*% fixed_effects

# Manually specify the coefficients and intercept for the line equation
slope_temp <- -0.05
slope_mass <- 0.19
intercept <- 4.17

# Generate a smooth grid for temperature
smooth_temp <- seq(min(iso_13C_filtered$temperature), max(iso_13C_filtered$temperature), length.out = 100)

# Calculate predictions using the specified formula
# Use the mean value of ln_body_mass for simplicity in this example
mean_mass <- mean(iso_13C_filtered$ln_body_mass)
predictions_manual <- slope_temp * smooth_temp + slope_mass * mean_mass + intercept

# Plot actual data points
plot(iso_13C_filtered$temperature, iso_13C_filtered$ln_turnover_time, 
     xlab = "Body Temperature (°C)", 
     ylab = bquote(ln(t[1/2]/m[.(0.19)]^{}~"[days/g"[.(0.19)]^{}~"]")),
     main = paste("Predicted ln(Turnover Time) Based on Lambda =", round(mean_lambda, 3)),
     col = "black", pch = 16)

# Add the manually computed line
lines(smooth_temp, predictions_manual, col = "#FF8247", lwd = 2)

# Add R² and equation to the plot in the lower left corner
text(x = min(iso_13C_filtered$temperature), y = min(iso_13C_filtered$ln_turnover_time) + 0.5,
     labels = bquote(R^2 == .(round(r_squared, 3))),
     pos = 4, cex = 0.8)

text(x = min(iso_13C_filtered$temperature), y = min(iso_13C_filtered$ln_turnover_time),
     labels = bquote("y = " ~ -0.05 * temp + 0.19 * mass + 4.17),
     pos = 4, cex = 0.8)
```

